### PR TITLE
Pr 2 of 2: Add Microarray Pathway Analysis - GSEA example

### PR DESCRIPTION
**Pull request by cbethell**
## Analysis Purpose

This PR closes issue #284

## Pull Request Stage

This is a **Refined PR** - needs review of details and polishing

The draft PR relevant to this PR is #339.

## Strategy

This PR iterates on open PR #345 by adding the GSEA and visualization steps to the microarray GSEA example analysis.

I copied over the steps from draft PR #339, incorporating the relevant feedback from there.

Therefore, this PR includes:
- The implementation of `clusterProfiler::GSEA()` -- per [this comment on the relevant draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#r513799939), some research into the function led me to the current implementation in this PR as it seems that `fgseaMultilevel()` is the recommended approach to use and the default for `clusterProfiler::GSEA()` when the `nPerm` argument is not provided (info gathered partly from the [fgsea docs](https://bioconductor.org/packages/devel/bioc/manuals/fgsea/man/fgsea.pdf) and partly from the warning message below):

Screen Shot 2020-10-29 at 1 09 40 PM

Also probably worth noting, when attempting to run `fgseaSimple()` and/or `fgseaMultilevel()` (without using the clusterProfiler wrapper), 0 results were returned (could be an error in my implementation however).

- To answer the [question re the pvalue cutoff](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#r513798718) on the draft PR, the p-value cutoff (FDR in this case) is an arbitrary cutoff which is now equal to *0.05* in this PR based on what seems to be _standard_ practice according to various docs (including but not limited to [Toward a gold standard for benchmarking gene set enrichment analysis](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbz158/5722384) and [clusterProfiler: an R Package for Comparing Biological Themes Among Gene Clusters](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3339379/)).

- The decision to create the pre-ranked gene vector based on the t-statistic rather than log fold change was explored based on [this comment from the draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339#discussion_r513817275) and eventually made based on what is recommended in [Discovering statistically significant pathways in expression profiling studies](https://www.pnas.org/content/pnas/102/38/13544.full.pdf) and the explanation from a [biostars forum](https://www.biostars.org/p/200715/) which says "Try to understand how they relate to the question you're interested in e.g. if you're most interested in effect size then the fold change is what you should use but if you're more interested in statistical significance then look for one of the statistics taking into consideration the assumptions they make e.g. t-test" as I believe we are encouraging users to look at statistical significance here.

- Visualizations of the GSEA results using `enrichplot::gseaplot()` was included in this PR as well as they were in the draft PR

- Note that I also set the seed in this PR as I noticed that the results were slightly changing and realized that I did indeed this important step.

## Concerns/Questions for reviewers:

Note that at line 317, I left a prompt to `START REVIEW HERE` as this is the point where PR 1 of 2 (#345) ends and the new content being added in this PR begins (I will remove the prompt once #345 is merged).

Overall, do the decisions included in this PR make sense?

Does the context re `GSEA()` (the goal of it and the implementation) make sense?
Do the prompts seem sufficient for users?

### Content checks
* [x] All `{{BLANKS}}` have been replaced with the correct content.
* [x] [Sources are cited](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#citing-sources-in-text)
* [x] Seed is set (if applicable)

### Formatting Checks
* [x] Removed any manual numbering of sections.
* [x] Removed any instances of chunk naming.
* [x] Comments and documentation are up to date.
* [x] All links have been checked and are properly formatted.

### Add datasets to S3
* [x] Added [data and metadata files to S3](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#adding-datasets-to-the-S3-bucket).

### Docker/Snakemake rendering components
* [ ] Added the `.html` [link to the navigation bar](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#add-new-analyses-to-the-navbar). (Done in PR 1 of 2 #345)
* [x] Any not yet added packages needed for this analysis have been added to the Dockerfile and it successfully builds.

### CODE REVIEW

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:334]:
```suggestion
This is step 1 -- create a gene list with statistics that GSEA will rank by.
```

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:338]:
```suggestion
# already sorted in the t-statistic values in descending order in the previous
```

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:353]:
```suggestion
The enrichment score (ES) is a pathway-level statistic, calculated using our gene-level statistics.
```

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:355]:
```suggestion
An ES is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
```
What does it mean to `start with the most highly ranked genes`?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:356]: Are you saying that ESs can be negative or positive where 0 indicates randomness? 

I think you could move this sentence up above the 355 line. Tell us intuitively what the score means and then if needed, we can get into how that's calculated.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:357]: This tidbit might be better absorbed when we look at the pvalues at 386 - 387? What do you think?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:409]: Might be a good place to use `top_n()`? We don't really need everything printed out per se. Up to you.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:422]: I like this sentence, can we add a sentence like it for describing the bottom panel? Aka what's the red dashed line about (looks like its the peak or max deviation from 0) we should probably highlight that and what it means.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:355]: I'm wondering if some of this information about _how_ the scores are calculated would be better absorbed if it were given along side the enrichment plots. 

(But the overview of GSEA bits should still be kept here I think).

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:430]: In case people run things out of order (which I think the type of user who are following these examples may be prone to do that) perhaps we should assign the plots and supply them directly to the `plot` instead of using `last_plot()`?

**Review by cansavvy — COMMENTED** [review]
I think this looks great! I just have some wording suggestions. 

I floated an idea about moving some of the "how this is calculated" info close to where the plots are since I think the plots illustrate that well and visuals can be helpful for absorbing th ideas. Let me know what you think.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:352]: Per @jaclyn-taroni's [comment on PR #345](https://github.com/AlexsLemonade/refinebio-examples/pull/345#discussion_r518109401), we will likely want to  move the explanation below up before the `gene ID conversion` section and reiterate here since we now need to explain some scientific decision making in the `gene ID conversion` section.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: Agreed 👍

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: I tried rephrasing this in the most recent changes, please let me know if it is still unclear.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: > What does it mean to start with the most highly ranked genes?

The most highly ranked genes here refer to the highest positive and highest negative ranked genes according the t-statistic (tried to clarify this in the latest updates, please let me know if it is still not clear).

> I'm wondering if some of this information about how the scores are calculated would be better absorbed if it were given along side the enrichment plots.

> (But the overview of GSEA bits should still be kept here I think).

I also tried to rearrange this section as you suggested here in the latest updates to this PR.

**Comment by cbethell** [comment]
@cansavvy, I believe I mostly addressed your review suggestions. Please let me know if anything is still not clear or if I misinterpreted any of your suggestions!

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:418]: Can print out more than 1 just to show a little bit of a preview? Not 100% necessary but might be nice.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:437]:
```suggestion
In this case, the most highly positive enrichment score's data are being displayed.
```
The graph is only associated with one NES right?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:457]: You can use -1 for this purpose instead of `desc()` That's what the docs show: https://dplyr.tidyverse.org/reference/top_n.html

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:460]:
```suggestion
The gene set `HALLMARK_E2F_TARGETS` has the most negative NES.
```

It didn't occur to me before but "highest negative" sounds weird. Should we change this or is it not as weird as I'm thinking it is?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:474]: This would probably be
```suggestion
Again, the red dashed line here indicates the maximum deviation from zero, in other words, the most negative ES score.
```

**Review by cansavvy — COMMENTED** [review]
This looks good. I mostly have wording comments and little comments.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: You're correct, this snuck back in from what was copied from training modules (got rid of it in the last commit).

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: I attempted to address this comment in the last commit via some restructuring and rewording. Please let me know if what I pushed in the latest commits are suffice (I was not sure if I should go more in depth about the warning message since users will not actually see it here).

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: Agreed, for consistency, I mimicked this language when speaking about the _most_ positive NES as well. Please let me know if you think I should change that back to _highest positive_.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:341]: This sentence and 339 seem too similarly worded to be helpful. I think repetition or reiteration of an important point is good, but it feels a little too echoey. Can you explain it a different way or probably just delete it

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:476]: The sentence here looks like its still describing positive ES. Feel free to change around this wording suggestion, but we need it to describe NES. 
```suggestion
Here the ES is calculated by starting with the most lowly ranked genes (according to the gene-level t-statistic values) and making the score more negative when a gene is in the pathway. 
```

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cansavvy) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:353]:
```suggestion
# We need to sort the t-statistic values in descending order here
t_vector <- sort(t_vector, decreasing = TRUE)
```
And then delete the second half of the comment above. I think placing the comment with its corresponding code makes it easier for the user to parse out which step is which.

**Review by cansavvy — COMMENTED** [review]
I have a couple wording requests but otherwise this looks good to me. I don't feel the need to review this again.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:340]: I don't think we _expect_ the latter distribution. It depends on the experiment! We might not always expect a particular pathway to be perturbed between our two groups. The way we run GSEA here (and almost every other way folks use GSEA) is a comparison between groups.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:339]: Is this the goal of the analysis or is it a description of the approach we use in service of our goal?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:395]: I _believe_ that this implementation permutes the gene labels and uses the same gene set? It's basically accomplishing the same thing – you get an distribution of ES values for a gene sets that are the same size as the _real_ gene set but the relationship between the gene set and the true ranking is "broken."  I would describe it the same way that the docs or book for clusterProfiler do. (I think this may come directly from training material and I made it a point to talk about there would be two different ways to accomplish the goal / I just wanted the point to land using whatever mental model worked best for someone.) I may be overthinking it. I can't tell.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:364]: I think we do want to move this up, possibly before we install the libraries or under `Purpose of this analysis`. I would probably collect some of the information from throughout what is being added here (this section, around line 280, around line 395) and write up something pretty comprehensive as an introduction to the method to start. Then I'd _consider_ paring back what is written alongside the code chunks to reiterate the most salient points.

**Review by jaclyn-taroni — COMMENTED** [review]
This looks like a great start! I had a few comments about wording and rearranging. I haven't looked at the code itself carefully or in great detail, but I assume that was a large part of @cansavvy's review at this point.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:339]: Good question, this goal was inspired by the goal in the related GSEA article: https://www.pnas.org/content/102/43/15545 which states that
"The goal of GSEA is to determine whether members of a gene set S tend to occur toward the top (or bottom) of the list L, in which case the gene set is correlated with the phenotypic class distinction."

However, now that you point it out, I would classify this as a description of the approach we use in service of our goal. Will update appropriately 👍

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:340]: Ahh okay that makes sense, will update this section to more accurately reflect what's going on here.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:395]: Gotcha! I'll revisit the docs and be sure to describe this section in a way which is more closely aligned with what's in the docs relevant to the package version we are using.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:347]: I feel like this may be leftover from training material where we say more about steps? Or maybe the training material doesn't say anything about steps after 1 #itwasapilot.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:438]: I _believe_ this is the first time we talk about this which seems _odd_. There may be something that isn't in the diff/was added earlier that I'm missing.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (jaclyn-taroni) [inline 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:478]: Do you have documentation to support this? I would imagine we wouldn't be able to tell which "end" to start with and by virtue of not having many of the genes at the beginning of the list you'd decrement the score a lot and therefore end up with a negative enrichment score.

**Review by jaclyn-taroni — APPROVED** [review]
This is in good shape and I am going to approve this. There's one comment that needs to be addressed around the negative enrichment scores before we merge this. Since we have to do #349, I think we should anticipate that some of what is currently written here will get rearranged and that means we should re-evaluate the text a bit when we get there. I see no reason to hold this up for that since it's hitting `staging` (woo the system is working)! Side note: can you file an issue to track updating this notebook specifically please?

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: I believe when I addressed one of the earlier review comments, I removed it from an earlier section but have now included it in the brief intro under the `Purpose of this analysis` section.

File: 02-microarray/pathway-analysis_microarray_03_gsea.Rmd
Comment (cbethell) [reply 02-microarray/pathway-analysis_microarray_03_gsea.Rmd]: Good question, this language was suggested in an earlier review comment but without the documentation to support it. I fixed the language in accordance with your comment above.

**Comment by cbethell** [comment]
> This is in good shape and I am going to approve this. There's one comment that needs to be addressed around the negative enrichment scores before we merge this. Since we have to do #349, I think we should anticipate that some of what is currently written here will get rearranged and that means we should re-evaluate the text a bit when we get there. I see no reason to hold this up for that since it's hitting `staging` (woo the system is working)! Side note: can you file an issue to track updating this notebook specifically please?

Addressed your comments in the [last commit](https://github.com/AlexsLemonade/refinebio-examples/pull/347/commits/82a97d882c72cff2d3ed67ee90e132aa69a48cd6) @jaclyn-taroni and filed issue #354 to track the context updates that we will need to make to the notebook in this PR.

**Comment by jaclyn-taroni** [comment]
I approved earlier, so this is fine to merge in my opinion! We will have to re-evaluate when we get to #354 (thanks for filing) to make sure our explanations of in this notebook are well-aligned with the intro, but this is good to go.

**Pull request merged**
